### PR TITLE
Add cache to space/{UUID}/developers calls

### PR DESCRIPTION
**Pull request by VaskoBozhurski**
#### Description: 

This commit adds a cache that stores the result per spaceGuid for the list
of developers in the AuthenticationChecker to reduce the number of calls
done to the controller.

If the user is not found in the list of developers, the cache will be
force refreshed to get fresh data from the controller, regardless of
expiration time.

#### Issue:
LMCROSSITXSADEPLOY-1429

### CODE REVIEW

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/CachedObject.java [hunk 1]
```diff
@@ -19,15 +19,25 @@ public CachedObject(long expirationTimeInSeconds, Supplier<Long> currentTimeSupp
...
 
-    public synchronized T get(Supplier<T> refreshFunction) {
-        long currentTime = currentTimeSupplier.get();
-        long millisecondsSinceLastRefresh = currentTime - lastRefreshTime;
-        long secondsSinceLastRefresh = TimeUnit.MILLISECONDS.toSeconds(millisecondsSinceLastRefresh);
-        if (object == null || secondsSinceLastRefresh > expirationTimeInSeconds) {
-            this.object = refreshFunction.get();
-            this.lastRefreshTime = currentTime;
+    public T get(Supplier<T> refreshFunction) {
+        synchronized (this) {
```
Comment (ddonchev) [inline com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/CachedObject.java]: method level

File: com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java [hunk 2]
```diff
@@ -34,6 +36,8 @@
 public class AuthorizationChecker {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(AuthorizationChecker.class);
+    private static final int CACHE_TIME_IN_SECONDS = 20;
```
Comment (ddonchev) [inline com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java]: ApplicationConfiguration

File: com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java [hunk 3]
```diff
@@ -111,11 +115,17 @@ boolean checkPermissions(UserInfo userInfo, String spaceGuid, boolean readOnly)
     }
 
     private boolean hasPermissions(CloudControllerClientSupportingCustomUserIds client, String userId, UUID spaceGuid, boolean readOnly) {
-
-        if (client.getSpaceDeveloperIdsAsStrings(spaceGuid)
+        // Get the space developers for a space GUID with cache.
+        if (SPACE_DEVELOPERS_CACHE.get(spaceGuid, (() -> getSpaceDevelopers(client, spaceGuid)))
```
Comment (ddonchev) [inline com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java]: extract if expressions to previous statements or method calls

File: com.sap.cloud.lm.sl.cf.web/src/test/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationCheckerTest.java [hunk 4]
```diff
@@ -120,6 +127,46 @@ public void checkPermissionsWithExceptionTest2() {
...
+        Mockito.verify(client, Mockito.times(1))
+            .getSpaceDeveloperIdsAsStrings(Mockito.eq(UUID.fromString(THIRD_SPACE_ID)));
+
+        UserInfo negativeUser = new UserInfo("newId", "newUser", userInfo.getToken());
+        when(clientProvider.getControllerClient(negativeUser.getName())).thenReturn(client);
+
+        assertFalse(authorizationChecker.checkPermissions(negativeUser, THIRD_SPACE_ID, false));
+        Mockito.verify(client, Mockito.times(2))
+            .getSpaceDeveloperIdsAsStrings(Mockito.eq(UUID.fromString(THIRD_SPACE_ID)));
+    }
```
Comment (ddonchev) [inline com.sap.cloud.lm.sl.cf.web/src/test/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationCheckerTest.java]: verify that the cache get's updated

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/CachedObject.java (hunk 1 above)
Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/CachedObject.java]: Done :)

File: com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java (hunk 3 above)
Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java]: Done.

File: com.sap.cloud.lm.sl.cf.web/src/test/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationCheckerTest.java (hunk 4 above)
Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.web/src/test/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationCheckerTest.java]: Done

File: com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/CachedMap.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+public class CachedMap<K, V> {
+
+    private long expirationTimeInSeconds;
+    private Supplier<Long> currentTimeSupplier;
+
+    private Map<K, CachedObject<V>> referenceMap = Collections
+        .synchronizedMap(new ReferenceMap<>(ReferenceStrength.HARD, ReferenceStrength.SOFT));
+
+    public CachedMap(long expirationTimeInSeconds, Supplier<Long> currentTimeSupplier) {
```
[the remark this answers is not in the document]
Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.core/src/main/java/com/sap/cloud/lm/sl/cf/core/model/CachedMap.java]: True, removed :)

File: com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java (hunk 2 above)
Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java]: Done

File: com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java [hunk 6]
```diff
@@ -46,6 +49,19 @@ public AuthorizationChecker(CloudControllerClientProvider clientProvider, SpaceG
         this.clientProvider = clientProvider;
         this.spaceGetter = spaceGetter;
         this.applicationConfiguration = applicationConfiguration;
+        if (spaceDevelopersCache == null) {
+            initSpaceDevelopersCache();
+        }
+    }
+
+    private void initSpaceDevelopersCache() {
+        synchronized (AuthorizationChecker.class) {
```
Comment (nictas) [inline com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java]: The class is a Spring singleton, so why not make the cache non-static. That way you'll be able to just make this method "private synchronized void".

Comment (VaskoBozhurski) [reply com.sap.cloud.lm.sl.cf.web/src/main/java/com/sap/cloud/lm/sl/cf/web/security/AuthorizationChecker.java]: Fixed

**Pull request merged**